Addition for ui-improvement-na50r (Client)

na50r opened the pull request:
In this branch I made some occasionally fixups on backend that frontend required it to have. For example: The result page didn't render the profile pictures because the `PlayerGetDTO` that was received didn't contain the user that had those picture, so I added the user as an attribute to `PlayerGetDTO`.

- I added Users to the `PlayerGetDTO`
- I set `lobby.gameTime` to 0 whenever a game ends, similar to how lobby is also set to `PREGAME`, this ensures that the timer must be set anew each new game.

Tmmn commented on src/main/java/ch/uzh/ifi/hase/soprafs24/service/GameService.java:
```diff
@@ -91,27 +100,40 @@ private Game instantiateGame(GameMode gameMode) {
     }
 
+    Player getOwner(Lobby lobby){
+        Player owner = playerService.findPlayerByToken(lobby.getOwner().getToken());
+        return owner;
```
SonarLint: Immediately return this expression instead of assigning it to the temporary variable "owner".

Tmmn commented on src/main/java/ch/uzh/ifi/hase/soprafs24/service/GameService.java:
```diff
@@ -91,27 +100,40 @@ private Game instantiateGame(GameMode gameMode) {
                 }
+
+                if(!getOwner(lobby).getStatus().equals(PlayerStatus.PLAYING)){
+                    gameTimer.cancel();
```
have you tried the approach to store gameTimer objects in a HashMap as GameService attributes and cancel them in the play method further up in this class? Service classes should only be instantiated once on server startup and stay the same as long as the server is running, but I have not confirmed that by testing it out.

na50r replied:
I implemented it with the HashMap now, thanks for the suggestion!

Tmmn commented on src/test/java/ch/uzh/ifi/hase/soprafs24/service/GameServiceTest.java:
```diff
@@ -112,34 +112,34 @@ public void play_success() {
-        verify(messagingTemplateMock, timeout(1000 * 20).times(3)).convertAndSend(eq("/topic/lobbies/1234/game"), any(TimeDTO.class));
-        verify(testLobby, timeout(1000 * 30).atLeastOnce()).setStatus(LobbyStatus.PREGAME);
-    }
+//
```
I assume you are still working on this

na50r replied:
Yes, still work in progress

na50r commented:
Changs in [this commit](https://github.com/sopra-fs24-group-41/sopra-fs24-group-41-server/pull/199/commits/afb20149028cf7dfcbda3014b21e9d3d2d29c3fe)

- I implemented Timer using HashMap, this allows me to access the timer also when the win condition is reached and terminate it.
- I fixed the Lobby Status change by accessing the Lobby through a LobbyService method I inserted. It is okay for GameService to know about LobbyService. LobbyService uses repository methods.

Tmmn commented on src/main/java/ch/uzh/ifi/hase/soprafs24/service/GameService.java:
```diff
@@ -113,20 +129,29 @@ private Game instantiateGame(GameMode gameMode) {
 
             public void run() {
+                if (remainingTime <= 0) {
+                    gameTimer.cancel(); // Stop the timer when time's up
+                    timers.remove(lobby.getCode());
+                    lobbyService.setStatusGivenLobby(lobby, LobbyStatus.PREGAME);
+                    messagingTemplate.convertAndSend("/topic/lobbies/" + lobby.getCode() + "/game", new InstructionDTO(Instruction.STOP));
```
extract end game method and reuse it in play method
wrap this action in a transaction like it's done here: https://github.com/sopra-fs24-group-41/sopra-fs24-group-41-server/blob/9b67480767977936fcc27f14bcc6604170835530/src/main/java/ch/uzh/ifi/hase/soprafs24/service/LobbyService.java#L73

Tmmn commented on src/main/java/ch/uzh/ifi/hase/soprafs24/service/GameService.java:
```diff
@@ -93,7 +102,14 @@ public Word play(Player player, List<Word> words) {
         updatePlayerStatistics(player, result);
 
         if (game.winConditionReached(player)) {
+            if(timers.get(lobby.getCode()) != null){
+                timers.get(lobby.getCode()).cancel();
+                timers.remove(lobby.getCode());
+            }
+
             lobby.setStatus(LobbyStatus.PREGAME);
+            lobby.setGameTime(0);
+            messagingTemplate.convertAndSend("/topic/lobbies/" + lobby.getCode() + "/game", new InstructionDTO(Instruction.STOP));
             updateWinsAndLosses(player, lobby);
             messagingTemplate.convertAndSend(String.format(MESSAGE_LOBBY_GAME, lobby.getCode()), new InstructionDTO(Instruction.STOP));
         }
```
reuse end game method here

Tmmn commented on src/main/java/ch/uzh/ifi/hase/soprafs24/service/GameService.java:
```diff
@@ -93,7 +102,14 @@ public Word play(Player player, List<Word> words) {
+
             lobby.setStatus(LobbyStatus.PREGAME);
+            lobby.setGameTime(0);
+            messagingTemplate.convertAndSend("/topic/lobbies/" + lobby.getCode() + "/game", new InstructionDTO(Instruction.STOP));
             updateWinsAndLosses(player, lobby);
             messagingTemplate.convertAndSend(String.format(MESSAGE_LOBBY_GAME, lobby.getCode()), new InstructionDTO(Instruction.STOP));
```
avoid sending the stop instruction two times

Tmmn commented on src/main/java/ch/uzh/ifi/hase/soprafs24/service/LobbyService.java:
```diff
@@ -233,4 +233,13 @@ private long generateLobbyCode() {
         return code;
     }
+
+    public void setStatusGivenLobby(Lobby lobby, LobbyStatus givenStatus) {
+        Lobby foundLobby = lobbyRepository.findByCode(lobby.getCode());
+        if(foundLobby!=null){
+            foundLobby.setStatus(givenStatus);
+            lobbyRepository.saveAndFlush(foundLobby);
+        }
+    }
```
not needed when you execute your timer code in a transaction

Tmmn reviewed (approved):
no bugs found ❌🐞
